### PR TITLE
Adding the Bento Box macropad

### PR DESCRIPTION
**Pull request by Synth-and-Keys**
## Description

* keyboard files added
* default keymap included
* added to keyboards/synthandkeys

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* n/a

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/synthandkeys/bento_box/config.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2023 Synth-and-Keys (@Synth-and-Keys)
```
Comment (zvecr) [inline keyboards/synthandkeys/bento_box/config.h:1]: This file should be removed

File: keyboards/synthandkeys/bento_box/rules.mk [hunk 2]
```diff
@@ -0,0 +1,4 @@
+ENCODER_ENABLE = yes
+BOOTMAGIC_ENABLE = yes
+BOOTLOADER = stm32-dfu
+MCU = STM32F072
```
Comment (zvecr) [inline keyboards/synthandkeys/bento_box/rules.mk:4]:
```suggestion
# This file intentionally left blank
```

File: keyboards/synthandkeys/bento_box/readme.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+The Bento Box is a 15-key, 3-encoder macropad running on QMK.
+
+* Keyboard Maintainer: [Synth-and-Keys](https://github.com/Synth-and-Keys)
+* Hardware Supported: Bento Box PCB
+* Hardware Availability: https://www.etsy.com/shop/SynthandKeys
+
+Make example for this keyboard (after setting up your build environment):
+
+    make bento_box:default
```
Comment (zvecr) [inline keyboards/synthandkeys/bento_box/readme.md:13]:
```suggestion
    make synthandkeys/bento_box:default
```

File: keyboards/synthandkeys/bento_box/readme.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+Flashing example for this keyboard:
+
+    make bento_box:default:flash
```
Comment (zvecr) [inline keyboards/synthandkeys/bento_box/readme.md:17]:
```suggestion
    make synthandkeys/bento_box:default:flash
```

File: keyboards/synthandkeys/bento_box/info.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    },
+    "matrix_pins": {
+        "direct": [
+            [null, null, null, "A9", "A8"],
+            [null, "B10", null, "B9", "B6"],
+            ["B0","A5","A2", null, null],
+            ["A7","A4","A1","B8","B5"],
+            ["A6","A3","A0","B7","B4"]
+        ]
+    },
```
Comment (zvecr) [inline keyboards/synthandkeys/bento_box/info.json:22]:
```suggestion
    },
    "encoder": {
        "rotary": [
            { "pin_a": "B2", "pin_b": "B1" },
            { "pin_a": "A13", "pin_b": "A14" },
            { "pin_a": "A15", "pin_b": "B3" }
        ]
    },
    "bootmagic": {
        "matrix": [2, 0]
    },
```

File: keyboards/synthandkeys/bento_box/info.json [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    "keyboard_name": "bento_box",
+    "maintainer": "Synth-and-Keys",
+    "bootloader": "stm32-dfu",
+    "diode_direction": "COL2ROW",
+    "features": {
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": true
```
Comment (zvecr) [inline keyboards/synthandkeys/bento_box/info.json:12]:
```suggestion
        "bootmagic": true,
        "command": false,
        "console": false,
        "extrakey": true,
        "mousekey": true,
        "nkro": true
        "encoder": true
```

File: keyboards/synthandkeys/bento_box/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,66 @@
+#include QMK_KEYBOARD_H
```
Comment (zvecr) [inline keyboards/synthandkeys/bento_box/keymaps/default/keymap.c:1]:
```suggestion
// Copyright 2023 Synth-and-Keys (@Synth-and-Keys)
// SPDX-License-Identifier: GPL-2.0-or-later

#include QMK_KEYBOARD_H
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/synthandkeys/bento_box/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    else if (index == 2) {
+        if (clockwise) {
+            tap_code16(C(KC_Y));
+        } else {
+            tap_code16(C(KC_Z));
+        }
+    }
+    return true;
+}
+
```
Comment (drashna) [inline keyboards/synthandkeys/bento_box/keymaps/default/keymap.c:56]: This should be moved to the keyboard's c file ("bento_box.c"), that way the board has a default configuration, and can be overwritten.


```c
#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code(KC_VOLU);
        } else {
            tap_code(KC_VOLD);
        }
    }
/* TOP LEFT*/
    else if (index == 1) {
        if (clockwise) {
            tap_code16(C(KC_EQUAL));
        } else {
            tap_code16(C(KC_MINUS));
        }
    }
/* TOP RIGHT*/
    else if (index == 2) {
        if (clockwise) {
            tap_code16(C(KC_Y));
        } else {
            tap_code16(C(KC_Z));
        }
    }
    return true;
}
#endif
```

File: keyboards/synthandkeys/bento_box/bento_box.c [hunk 9]
```diff
@@ -0,0 +1,30 @@
+#ifdef ENCODER_ENABLE
```
Comment (zvecr) [inline keyboards/synthandkeys/bento_box/bento_box.c:1]:
```suggestion
// Copyright 2023 Synth-and-Keys (@Synth-and-Keys)
// SPDX-License-Identifier: GPL-2.0-or-later

#include "quantum.h"

#ifdef ENCODER_ENABLE
```

**Comment by drashna** [comment]
Lint has some complaints that should be addressed

**Comment by Synth-and-Keys** [comment]
Resolved lint complaints. It says there are changes requested, but I believe I have already resolved all of them?

**Review by Synth-and-Keys — COMMENTED** [review]
Just trying to close out the change requests to move forward

**Review by drashna — APPROVED** [review]

File: keyboards/synthandkeys/bento_box/info.json [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                { "matrix": [3, 0], "x": 0, "y": 3 },
+                { "matrix": [3, 1], "x": 1, "y": 3 },
+                { "matrix": [3, 2], "x": 2, "y": 3 },
+                { "matrix": [3, 3], "x": 3.5, "y": 2.5 },
+                { "matrix": [3, 4], "x": 4.5, "y": 2.5 },
+                { "matrix": [4, 0], "x": 0, "y": 4 },
+                { "matrix": [4, 1], "x": 1, "y": 4 },
+                { "matrix": [4, 2], "x": 2, "y": 4 },
+                { "h": 1.5, "matrix": [4, 3], "x": 3.5, "y": 3.5 },
+                { "h": 1.5, "matrix": [4, 4], "x": 4.5, "y": 3.5 }
```
Comment (fauxpark) [inline keyboards/synthandkeys/bento_box/info.json:71]:
```suggestion
                {"matrix": [0, 3], "x": 3.5, "y": 0},
                {"matrix": [0, 4], "x": 4.5, "y": 0},
                {"matrix": [1, 1], "x": 1, "y": 0.5},
                {"matrix": [1, 3], "x": 3.5, "y": 1},
                {"matrix": [1, 4], "x": 4.5, "y": 1},
                {"matrix": [2, 0], "x": 0, "y": 2},
                {"matrix": [2, 1], "x": 1, "y": 2},
                {"matrix": [2, 2], "x": 2, "y": 2},
                {"matrix": [3, 0], "x": 0, "y": 3},
                {"matrix": [3, 1], "x": 1, "y": 3},
                {"matrix": [3, 2], "x": 2, "y": 3},
                {"matrix": [3, 3], "x": 3.5, "y": 2.5},
                {"matrix": [3, 4], "x": 4.5, "y": 2.5},
                {"matrix": [4, 0], "x": 0, "y": 4},
                {"matrix": [4, 1], "x": 1, "y": 4},
                {"matrix": [4, 2], "x": 2, "y": 4},
                {"matrix": [4, 3], "x": 3.5, "y": 3.5, "h": 1.5},
                {"matrix": [4, 4], "x": 4.5, "y": 3.5, "h": 1.5}
```

File: keyboards/synthandkeys/bento_box/info.json [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+                "pin_b": "B1"
+            },
+            {
+                "pin_a": "A13",
+                "pin_b": "A14"
+            },
+            {
+                "pin_a": "A15",
+                "pin_b": "B3"
+            }
```
Comment (fauxpark) [inline keyboards/synthandkeys/bento_box/info.json:23]:
```suggestion
            {"pin_a": "B2", "pin_b": "B1"},
            {"pin_a": "A13", "pin_b": "A14"},
            {"pin_a": "A15", "pin_b": "B3"}
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

File: keyboards/synthandkeys/bento_box/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    * ----------------------------------------------------------------
+    * |  Left  |  Down   |  Right  |  Shift+Alt+Tab  |    Alt+Tab    |
+    * ----------------------------------------------------------------
+     */
+    [0] = LAYOUT(
+                                             (C(KC_0)),  (G(S(KC_S))),
+                   KC_MUTE,                  (C(KC_C)),     (C(KC_V)),
+        KC_MPRV,   KC_MPLY,    KC_MNXT,
+        KC_HOME,     KC_UP,     KC_END,      (C(KC_T)),     (C(KC_W)),
+        KC_LEFT,   KC_DOWN,   KC_RIGHT, (S(C(KC_TAB))),   (C(KC_TAB))
```
Comment (fauxpark) [inline keyboards/synthandkeys/bento_box/keymaps/default/keymap.c:25]:
```suggestion
                                             C(KC_0),  G(S(KC_S)),
                   KC_MUTE,                  C(KC_C),     C(KC_V),
        KC_MPRV,   KC_MPLY,    KC_MNXT,
        KC_HOME,     KC_UP,     KC_END,      C(KC_T),     C(KC_W),
        KC_LEFT,   KC_DOWN,   KC_RIGHT, S(C(KC_TAB)),   C(KC_TAB)
```

File: keyboards/synthandkeys/bento_box/info.json [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        "command": false,
+        "console": false,
+        "encoder": true,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": true
+    },
+    "matrix_pins": {
+        "direct": [
+            [null, null, null, "A9", "A8"],
```
Comment (fauxpark) [inline keyboards/synthandkeys/bento_box/info.json:28]: You should not have any null pins here.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/synthandkeys/bento_box/info.json [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+        "mousekey": true,
+        "nkro": true
+    },
+    "matrix_pins": {
+        "direct": [
+            ["A9", "A8"],
+            ["B9", "B6"],
+            ["B0", "A5", "A2"],
+            ["A7", "A4", "A1", "B8", "B5"],
+            ["A6", "A3", "A0", "B7", "B4"]
```
Comment (fauxpark) [inline keyboards/synthandkeys/bento_box/info.json:32]: You can make better use of the matrix like so:
```suggestion
            ["A9", "A8", "B9", "B6", "B0", "A5"],
            ["A2", "A7", "A4", "A1", "B8", "B5"],
            ["A6", "A3", "A0", "B7", "B4", null]
```
Then adjust the layout matrix positions appropriately.

File: keyboards/synthandkeys/bento_box/info.json [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+                { "matrix": [3, 0], "x": 0, "y": 3 },
+                { "matrix": [3, 1], "x": 1, "y": 3 },
+                { "matrix": [3, 2], "x": 2, "y": 3 },
+                { "matrix": [3, 3], "x": 3.5, "y": 2.5 },
+                { "matrix": [3, 4], "x": 4.5, "y": 2.5 },
+                { "matrix": [4, 0], "x": 0, "y": 4 },
+                { "matrix": [4, 1], "x": 1, "y": 4 },
+                { "matrix": [4, 2], "x": 2, "y": 4 },
+                { "matrix": [4, 3], "x": 3.5, "y": 3.5, "h": 1.5},
+                { "matrix": [4, 4], "x": 4.5, "y": 3.5, "h": 1.5}
```
Comment (fauxpark) [inline keyboards/synthandkeys/bento_box/info.json:60]:
```suggestion
                { "matrix": [0, 0], "x": 3.5, "y": 0 },
                { "matrix": [0, 1], "x": 4.5, "y": 0 },

                { "matrix": [0, 2], "x": 1, "y": 0.5 },
                { "matrix": [0, 3], "x": 3.5, "y": 1 },
                { "matrix": [0, 4], "x": 4.5, "y": 1 },

                { "matrix": [0, 5], "x": 0, "y": 2 },
                { "matrix": [1, 0], "x": 1, "y": 2 },
                { "matrix": [1, 1], "x": 2, "y": 2 },

                { "matrix": [1, 2], "x": 0, "y": 3 },
                { "matrix": [1, 3], "x": 1, "y": 3 },
                { "matrix": [1, 4], "x": 2, "y": 3 },
                { "matrix": [1, 5], "x": 3.5, "y": 2.5 },
                { "matrix": [2, 0], "x": 4.5, "y": 2.5 },

                { "matrix": [2, 1], "x": 0, "y": 4 },
                { "matrix": [2, 2], "x": 1, "y": 4 },
                { "matrix": [2, 3], "x": 2, "y": 4 },
                { "matrix": [2, 4], "x": 3.5, "y": 3.5, "h": 1.5},
                { "matrix": [2, 5], "x": 4.5, "y": 3.5, "h": 1.5}
```
This layout doesn't make sense; there are 18 keys but the direct_pins configuration above only specifies 17 pins (hence the single `null` pin). Please double check this.

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**